### PR TITLE
authorization: clarify/fix service accounts and workspace initialization

### PR DESCRIPTION
**Pull request by sttts**
Changes:
1. give initializers admin access
2. allow service account to access/admin sub-workspaces if permission is granted
3. fix some loop-holes for service account in the authz logic

TODOs:
- [ ] add unit and e2e tests

### CODE REVIEW

File: docs/authorization.md [hunk 1]
```diff
@@ -154,6 +155,29 @@ roleRef:
   name: workspace-admin
 ```
 
+### Initializing Workspaces
+
+By default, workspaces are only accessible to a user if they are in phse `Ready`. Workspaces that are initializing
+can be access only by users that are granted `initialize` verb on the `clusterworkspaces/content` resource in the
```
Comment (stevekuznetsov) [inline docs/authorization.md:161]:
```suggestion
can be accessed only by users that are granted the `initialize` verb on the `clusterworkspaces/content` resource in the 
```

File: docs/authorization.md [hunk 2]
```diff
@@ -154,6 +155,29 @@ roleRef:
   name: workspace-admin
 ```
 
+### Initializing Workspaces
+
+By default, workspaces are only accessible to a user if they are in phse `Ready`. Workspaces that are initializing
```
Comment (stevekuznetsov) [inline docs/authorization.md:160]:
```suggestion
By default, workspaces are only accessible to a user if they are in phase `Ready`. Workspaces that are initializing
```

File: docs/authorization.md [hunk 3]
```diff
@@ -154,6 +155,29 @@ roleRef:
...
+can be access only by users that are granted `initialize` verb on the `clusterworkspaces/content` resource in the 
+parent workspace.
+
+E.g. the workspace `root:org:ws:ws` initializing user is bound to a role of the following shape in `root:org:ws`:
```
Comment (stevekuznetsov) [inline docs/authorization.md:164]:
```suggestion
E.g. the actor that initializes the workspace `root:org:ws:ws` is bound to a role of the following shape in `root:org:ws`:
```

File: docs/authorization.md [hunk 4]
```diff
@@ -170,3 +194,13 @@ and work just like in a regular Kubernetes cluster.
 Note: groups added by the workspace content authorizer can be used for role bindings in that workspace.
 
 It is possible to bind to roles and cluster roles in the bootstrap policy from a local policy `RoleBinding` or `ClusterRoleBinding`.
+
+# Service Accounts
+
+Kubernetes service accounts is granted access to the workspaces they are defined in and that are ready.
```
Comment (stevekuznetsov) [inline docs/authorization.md:200]:
```suggestion
Kubernetes service accounts are granted access to the workspaces they are defined when the workspace is in the `Ready` phase. 
```

File: docs/authorization.md [hunk 5]
```diff
@@ -170,3 +194,13 @@ and work just like in a regular Kubernetes cluster.
 Note: groups added by the workspace content authorizer can be used for role bindings in that workspace.
 
 It is possible to bind to roles and cluster roles in the bootstrap policy from a local policy `RoleBinding` or `ClusterRoleBinding`.
+
+# Service Accounts
+
+Kubernetes service accounts is granted access to the workspaces they are defined in and that are ready. 
+
+E.g. a service account "default" in `root:org:ws:ws` is granted access to `root:org:ws:ws`, and through the
```
Comment (stevekuznetsov) [inline docs/authorization.md:202]: Need to be more clear about the `:ws:ws` nesting here

Comment (stevekuznetsov) [reply docs/authorization.md:202]: What does "granted access" mean?

File: docs/authorization.md [hunk 6]
```diff
@@ -170,3 +194,13 @@ and work just like in a regular Kubernetes cluster.
...
+workspace content authorizer it gains the `system:kcp:clusterworkspace:access` group membership.
```
Comment (stevekuznetsov) [inline docs/authorization.md:203]: Is this what it gains only during `Ready` phase?

File: docs/authorization.md [hunk 7]
```diff
@@ -170,3 +194,13 @@ and work just like in a regular Kubernetes cluster.
...
+
+A service account can gain access (both pure access and `admin` access) to sub-workspaces defined in the workspace
```
Comment (stevekuznetsov) [inline docs/authorization.md:205]: How is the distinction between "pure" and "admin" important?

File: pkg/authorization/toplevel_org_authorizer.go [hunk 8]
```diff
@@ -70,40 +70,54 @@ func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr author
...
 
-	// everybody authenticated has access to the root workspace
-	if cluster.Name == v1alpha1.RootCluster {
-		if sets.NewString(attr.GetUser().GetGroups()...).Has("system:authenticated") {
+	subjectClusters := map[logicalcluster.LogicalCluster]bool{}
+	for _, sc := range attr.GetUser().GetExtra()[authserviceaccount.ClusterNameKey] {
+		subjectClusters[logicalcluster.New(sc)] = true
+	}
+
+	// everybody authenticated has access to the root workspace, for service account only when there defined
```
Comment (stevekuznetsov) [inline pkg/authorization/toplevel_org_authorizer.go:83]: > for service account only when there defined

What does this mean?

File: pkg/authorization/workspace_content_authorizer.go [hunk 9]
```diff
@@ -100,91 +101,89 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
-			return authorizer.DecisionNoOpinion, "", err
-		} else if len(ws.Status.Initializers) > 0 {
+	extraGroups := sets.NewString()
+
+	// check the workspace even exists
+	if ws, err := a.clusterWorkspaceLister.Get(clusters.ToClusterAwareKey(parentClusterName, clusterWorkspace)); err != nil {
+		if errors.IsNotFound(err) {
+			return authorizer.DecisionDeny, fmt.Sprintf("%q workspace access not permitted", cluster.Name), nil
+		}
+		return authorizer.DecisionNoOpinion, "", err
```
Comment (stevekuznetsov) [inline pkg/authorization/workspace_content_authorizer.go:111]: Is this the sort of place we want to bubble up infra level issues to the user? Why not a generic 500?

File: pkg/authorization/workspace_content_authorizer.go [hunk 10]
```diff
@@ -100,91 +101,89 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
+	} else if len(ws.Status.Initializers) > 0 {
+		// service accounts can potentially access sub-workspaces, but not other workspaces during initialization. Hence, these two case:
+		// 1. it's  not a serivce account
```
Comment (stevekuznetsov) [inline pkg/authorization/workspace_content_authorizer.go:114]:
```suggestion
		// 1. it's  not a service account
```

File: pkg/authorization/workspace_content_authorizer.go [hunk 11]
```diff
@@ -74,23 +74,24 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
 	}
-	if cluster == nil || cluster.Name.Empty() {
+	if cluster == nil || cluster.Name.Empty() || !cluster.Name.HasPrefix(v1alpha1.RootCluster) {
 		return authorizer.DecisionNoOpinion, fmt.Sprintf("%q workspace access not permitted", cluster.Name), nil
 	}
 
-	// everybody authenticated has access to the root workspace
+	clusterWorkspace := cluster.Name.Base()
+	subjectClusters := map[logicalcluster.LogicalCluster]bool{}
+	for _, sc := range attr.GetUser().GetExtra()[authserviceaccount.ClusterNameKey] {
```
Comment (stevekuznetsov) [inline pkg/authorization/workspace_content_authorizer.go:83]: nit: this likely could use a helper to DRY attr -> cluster

File: pkg/authorization/workspace_content_authorizer.go [hunk 12]
```diff
@@ -100,91 +101,89 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
+		return authorizer.DecisionNoOpinion, "", err
+	} else if len(ws.Status.Initializers) > 0 {
+		// service accounts can potentially access sub-workspaces, but not other workspaces during initialization. Hence, these two case:
+		// 1. it's  not a serivce account
+		// 2. it's a service account, and it is defined in the parent workspace
+		if len(subjectClusters) == 0 && subjectClusters[parentClusterName] {
 			workspaceAttr := authorizer.AttributesRecord{
 				User:            attr.GetUser(),
-				Verb:            attr.GetVerb(),
+				Verb:            "initialize",
```
Comment (stevekuznetsov) [inline pkg/authorization/workspace_content_authorizer.go:119]: what is the axiom by which we evaluate the previous (subresource=initialize) to this proposed (verb=initialize) approach?

**Review by stevekuznetsov — COMMENTED** [review]
leaving some comments as I try to grok this

File: pkg/authorization/workspace_content_authorizer.go [hunk 13]
```diff
@@ -100,91 +101,89 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
+	if ws, err := a.clusterWorkspaceLister.Get(clusters.ToClusterAwareKey(parentClusterName, clusterWorkspace)); err != nil {
+		if errors.IsNotFound(err) {
+			return authorizer.DecisionDeny, fmt.Sprintf("%q workspace access not permitted", cluster.Name), nil
+		}
+		return authorizer.DecisionNoOpinion, "", err
+	} else if len(ws.Status.Initializers) > 0 {
+		// service accounts can potentially access sub-workspaces, but not other workspaces during initialization. Hence, these two case:
+		// 1. it's  not a serivce account
+		// 2. it's a service account, and it is defined in the parent workspace
+		if len(subjectClusters) == 0 && subjectClusters[parentClusterName] {
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:116]: @s-urbaniak `if len(subjectClusters) > 0`

Comment (s-urbaniak) [reply pkg/authorization/workspace_content_authorizer.go:116]: @s-urbaniak mind, `subjectClusters` is set in the SA's JWTs

File: pkg/authorization/workspace_content_authorizer.go [hunk 14]
```diff
@@ -100,91 +101,89 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
-				User:            attr.GetUser(),
-				Verb:            verb,
-				APIGroup:        v1alpha1.SchemeGroupVersion.Group,
-				APIVersion:      v1alpha1.SchemeGroupVersion.Version,
-				Resource:        "clusterworkspaces",
-				Subresource:     "content",
-				Name:            clusterWorkspace,
-				ResourceRequest: true,
+		// service accounts can get access or be admin in sub-workspaces, but not other workspaces
+		if len(subjectClusters) == 0 && subjectClusters[parentClusterName] {
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:144]: @s-urbaniak `len(subjectClusters) > 0`

File: pkg/authorization/workspace_content_authorizer.go [hunk 15]
```diff
@@ -100,91 +101,89 @@ func (a *workspaceContentAuthorizer) Authorize(ctx context.Context, attr authori
...
-		// a subject from a workspace, like a ServiceAccount, is automatically authenticated
-		// against that workspace.
-		// On the other hand, referencing that in the parent cluster for further permissions
-		// is not possible. Hence, we skip the authorization steps for the verb below.
-		for _, sc := range subjectCluster {
-			if logicalcluster.New(sc) == cluster.Name {
-				extraGroups = append(extraGroups, bootstrap.SystemKcpClusterWorkspaceAccessGroup)
-				break
+			if dec == authorizer.DecisionAllow {
+				extraGroups.Insert(bootstrap.SystemKcpClusterWorkspaceAccessGroup)
```
Comment (s-urbaniak) [inline pkg/authorization/workspace_content_authorizer.go:132]: additionally, add `SystemKcpClusterWorkspaceAdminGroup`. Reasoning: if the SA has `initialize` permissions on the parent ws, it is expected that this SA has administrative permissions on sub-workspaces that have initializers declared.

The initializing controller (the one who has a SA on the parent ws with `initialize` verbs) thus has the same permissions as an `admin` user manually initializing a universal sub-workspace.

**Pull request closed without merging**